openhcl/virt_mshv_vtl/src/processor/tdx/mod.rs

```diff
@@ -2394,6 +2369,93 @@ impl UhProcessor<'_, TdxBacked> {
         self.backing.vtls[vtl].exception_error_code = 0;
     }
 
+    fn inject_mc(&mut self, vtl: GuestVtl) {
+        self.backing.vtls[vtl].interruption_information = InterruptionInformation::new()
+            .with_valid(true)
+            .with_vector(x86defs::Exception::MACHINE_CHECK.0)
+            .with_interruption_type(INTERRUPT_TYPE_HARDWARE_EXCEPTION);
+    }
+
+    async fn handle_ept(
+        &mut self,
+        intercepted_vtl: GuestVtl,
+        dev: &impl CpuIo,
+        gpa: u64,
+        ept_info: VmxEptExitQualification,
+    ) -> Result<(), VpHaltReason<UhRunVpError>> {
+        let vtom = self.partition.caps.vtom.expect("vtom always set on tdx");
+        let is_shared = (gpa & vtom) == vtom;
+        let canonical_gpa = gpa & !vtom;
+
+        // Only emulate the access if the gpa is expected to be accessible. This
+        // means, the gpa was described to VTL0 in some form as memory or mmio,
+        // and the hardware did not generate an exit for a private/shared
+        // violation.
+        let address_type = self
+            .partition
+            .lower_vtl_memory_layout
+            .probe_address(canonical_gpa);
+
+        match address_type {
```

Comment:
Probably this logic should ultimately move into the virt_support_x86emu crate so that we don't have to replicate it everywhere. I.e., we should probably do this same thing for standard VMs with openvmm, right?